File: Leetcode/Array&String/String/76_minimum-window-substring/main.cpp
// 76. 最小覆盖子串
// Created by bob on 2022/9/6.
//
#include<string>
#include<unordered_map>

using namespace std;

class Solution {
public:
    unordered_map<char, int> ori, cnt;

    bool check() {
        for (auto &p:ori) {
            if (cnt[p.first] < p.second) {
                return false;
            }
        }
        return true;
    }

    string minWindow(string s, string t) {
        for (auto &c: t) {
            ori[c]++;
        }

        int l = 0, r = -1;
        int len = INT_MAX, ansL = -1, ansR = -1;
        while (r < int(s.size())) {
            if (ori.find(s[++r]) != ori.end()) {
                cnt[s[r]]++;
            }
            while (check() && l <= r) {
                if (r - l + 1 < len) {
                    len = r- l + 1;
                    ansL = l;
                }
                if (ori.find(s[l]) != ori.end()) {
                    cnt[s[l]]--;
                }
                ++l;
            }
        }
        return ansL == -1 ? string() : s.substr(ansL, len);
    }
};

int main() {
    string s = "ADOBECODEBANC";
    string t = "ABC";
    string res = Solution().minWindow(s, t);
    printf("%s", res.c_str());
    return 0;
}

File: Leetcode/Array&String/String/76_minimum-window-substring/main.py
class Solution:

    def minWindow(self, s: str, t: str) -> str:
        ori = {}.fromkeys(set(t), 0)
        for ch in t:
            ori[ch] += 1

        cnt = {}.fromkeys(set(t), 0)

        def check():
            for k, v in ori.items():
                if cnt[k] < v:
                    return False
            return True

        l, r = 0, -1
        min_len = 99999
        ansL, ansR = -1, -1
        while r < len(s):
            r += 1
            if r < len(s) and s[r] in ori:
                cnt[s[r]] += 1

            while check() and l <= r:
                if r - l + 1 < min_len:
                    min_len = r - l + 1
                    ansL = l
                if s[l] in ori:
                    cnt[s[l]] -= 1
                l += 1
        return "" if ansL==-1 else s[ansL: ansL+min_len]


if __name__ == '__main__':
    s = "a"
    t = "aa"
    res = Solution().minWindow(s, t)
    print(res)
